### DISCUSSION TITLE
Variables in utilities possibly not working or explained vaguely

### DISCUSSION THREAD

**Discussion by arpadgabor** [windicss/windicss#45] [Q&A] [answered]
First of all, this project is amazing, very nice work!

So I'm playing around with utility classes and variables and I found out that the documentation is a bit vague in this regard (was not sure where to post because this might also be a bug.)

As per docs:

> You can even pass in variable names, which is very useful in combination with css variables.
> ```css
> bg-$${variableName}
> 
> .bg-$test-variable {
>   --tw-bg-opacity: 1;
>   background-color: rgba(var(--test-variable), var(--tw-bg-opacity));
> }
> ```

However, it is not made clear _how_ you defined these variables. Are they equivalent to custom css properties defined in `:root`? Do we use template literals? Are they somehow magically picked out from js? Can they even be used at runtime? Is this only for colors?

Thus, to discover this behavior I tried all methods that I thought of in Vue.

```html
<!-- marginTop declared as reactive state -->
<div class="mt-$${marginTop}" />

<!-- marginTop declared as reactive state and with template literals -->
<div :class="'mt-$${' + marginTop + '}'" />
<div :class="`mt-$${marginTop}`" />
<div :class="`mt-${marginTop}`" />
<div :class="'mt-${' + marginTop + '}'" />

<!-- margin-top as a css custom property declared in :root -->
<div :class="`mt-$${margin-top}`" />
<div :class="`mt-${margin-top}`" />

<!-- Admittedly this was what I tried in the last half an hour or so -->
```

But none of them actually worked. Maybe I'm understanding it wrong, and some clarifications are needed in the docs, or maybe it is indeed a bug.

Any thoughts, ideas or suggestions are welcome! Thanks.

**Comment by arpadgabor** [comment, marked as answer]
As per usual I find the solution after I post the problem.

```vue
<script>
  // [module.methods.addMargin]
  const vars = document.documentElement.style

  this.marginTop++
  vars.setProperty('--margin-top', `${this.marginTop}rem`)
</script>

<template>
  <div class="mt-$margin-top">
    Hello world!
  </div>
</template>
```

Though I guess making it explicit that you need to use CSS Custom Properties, not variables, as they may be confused, like I did. 🤦‍♂️

**Reply by voorjaar** [reply]
Okay, I will update the documentation on this.
